### ISSUE TITLE
Weird behavior of Photo.visible

### ISSUE THREAD

**Issue by cmrd-senya** [diaspora/diaspora#6811]
I was hunting down the issue that lead to the wrong photo number in the [tests](https://travis-ci.org/diaspora/diaspora/jobs/124726853#L561) of #6783. Basically, it happens that [this test](https://github.com/diaspora/diaspora/blob/develop/spec/controllers/photos_controller_spec.rb#L107) generates 16 test photos, but when it checks the number with the ```Photo.visible``` method, the value is bigger, e.g. 19.

Finally I managed to get some database dump that shows weird result and seemingly  reproduces that very bug. I worked with mysql, but it happens on postgres as well ([example](https://travis-ci.org/diaspora/diaspora/jobs/124726857#L573)).

I [rebuilt](https://gist.github.com/cmrd-senya/0683ab0c8c024620392a9e525ed5f025) the db state similar to what we have in the test. If you load this DB dump and run the following commands within the rails console, you'll see the issue:
```
2.2.1 :001 > alice = User.find_by(username: "alice")
 => #<User id: 9, username: "alice", serialized_private_key: "-----BEGIN RSA PRIVATE KEY-----\nMIICWwIBAAKBgQDTH4...", getting_started: false, disable_mail: false, language: "en", email: "bob151f078@pivotallabs.com", encrypted_password: "$2a$04$acc98iRbowJhXqnH3I1SNuLI.MaVUAdj2/MyRC40GVA...", invitation_token: nil, invitation_sent_at: nil, reset_password_token: nil, remember_created_at: nil, sign_in_count: 0, current_sign_in_at: nil, last_sign_in_at: nil, current_sign_in_ip: nil, last_sign_in_ip: nil, created_at: "2016-04-22 17:28:17", updated_at: "2016-04-22 17:28:17", invitation_service: nil, invitation_identifier: nil, invitation_limit: nil, invited_by_id: nil, invited_by_type: nil, authentication_token: nil, unconfirmed_email: nil, confirm_email_token: nil, locked_at: nil, show_community_spotlight_in_stream: true, auto_follow_back: false, auto_follow_back_aspect_id: nil, hidden_shareables: {}, reset_password_sent_at: nil, last_seen: nil, remove_after: nil, export: "alice_diaspora_data_i-YlfKsk9vQLKIhtUvYpIA.json.gz", exported_at: nil, exporting: false, strip_exif: false, exported_photos_file: nil, exported_photos_at: nil, exporting_photos: false, color_theme: "original"> 
2.2.1 :002 > eve = User.find_by(username: "eve")
 => #<User id: 10, username: "eve", serialized_private_key: "-----BEGIN RSA PRIVATE KEY-----\nMIICWwIBAAKBgQDTH4...", getting_started: false, disable_mail: false, language: "en", email: "bob20d25d6@pivotallabs.com", encrypted_password: "$2a$04$GZReLqLhTFkGu0/gDzLs1eCzH2RQj4Q.gpbDlnVABtx...", invitation_token: nil, invitation_sent_at: nil, reset_password_token: nil, remember_created_at: nil, sign_in_count: 0, current_sign_in_at: nil, last_sign_in_at: nil, current_sign_in_ip: nil, last_sign_in_ip: nil, created_at: "2016-04-22 17:28:17", updated_at: "2016-04-22 17:28:17", invitation_service: nil, invitation_identifier: nil, invitation_limit: nil, invited_by_id: nil, invited_by_type: nil, authentication_token: nil, unconfirmed_email: nil, confirm_email_token: nil, locked_at: nil, show_community_spotlight_in_stream: true, auto_follow_back: false, auto_follow_back_aspect_id: nil, hidden_shareables: {}, reset_password_sent_at: nil, last_seen: nil, remove_after: nil, export: "eve_diaspora_data_JVcwGMM0IJA2zsxQwEobrg.json.gz", exported_at: nil, exporting: false, strip_exif: true, exported_photos_file: nil, exported_photos_at: nil, exporting_photos: false, color_theme: "original"> 
2.2.1 :003 > Photo.count
 => 8 
2.2.1 :004 > Photo.visible(alice, eve.person).count(:all)
 => 9 
```

So, basically, the method returns the visible photos count bigger than the total photos count, which is, apparently, a bug.

I can't say how does it reflect on the application operation, but I catch it sometimes in the testsuite of my PR #6783. I don't know why it happens only sometimes but not everytime. I guess the underlying SQL query of ```Photo.visible``` must be studied and I'm a bit tired to do that right now.

**Comment by cmrd-senya** [comment]
Ok, here is what happens.

This is an example of SQL that runs when ```Photo.visible``` method is called.
```sql
SELECT COUNT(*) FROM `photos` LEFT OUTER JOIN share_visibilities ON share_visibilities.shareable_id = photos.id WHERE `photos`.`author_id` = 28 AND (`share_visibilities`.`user_id` = 9 AND `share_visibilities`.`shareable_type` = 'Photo' OR `photos`.`public` = 1) AND (photos.created_at < '2016-04-25 11:03:37.677620')
```

You may see, that JOIN is performed on IDs equality, but it doesn't consider ```shareable_type```. So if there is a Post with an ID equal to some photo ID, and if we have a ShareVisibility for that Post, then in this query the Post's SV will be joined with the Photo data. And if we have several of ShareVisibilies for the same ID, then it'll return several copies of the Photo row, each for every SV. There is the `share_visibilities`.`shareable_type` = 'Photo' restriction, but it is respected only when `photos`.`public` != 1 because of the `OR` condition.

Volatile reproducibility is explained by the fact, that the IDs of photos and posts are changed on every test run, because we regenerate the test data (and the DB counters aren't nullified).

I guess it can be fixed by moving  `share_visibilities`.`shareable_type` = 'Photo' condition up to the join condition.

I'm going to prepare a PR to fix this.
